pkg/operator/ceph/object/cosi/controller.go

```diff
@@ -0,0 +1,196 @@
+}
+
+// Reconcile reads that state of the cluster for a CephCOSIDriver object and makes changes based on the state read
+// and what is in the CephCOSIDriver.Spec
+func (r *ReconcileCephCOSIDriver) Reconcile(context context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reconcileResponse, cephCOSIDriver, err := r.reconcile(request)
+
+	return reporting.ReportReconcileResult(logger, r.recorder, request, &cephCOSIDriver, reconcileResponse, err)
+}
+
+func (r *ReconcileCephCOSIDriver) reconcile(request reconcile.Request) (reconcile.Result, cephv1.CephCOSIDriver, error) {
+	cephCOSIDriver := &cephv1.CephCOSIDriver{}
+
+	// Fetch the CephCOSIDriver instance
+	err := r.client.Get(r.opManagerContext, request.NamespacedName, cephCOSIDriver)
+	logger.Debugf("CephCOSIDriver: %+v", cephCOSIDriver)
+	if err != nil && client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to get CephCOSIDriver. %+v", err)
+	}
+
+	cosiDeploymentStrategy := "auto"
+	if !reflect.DeepEqual(cephCOSIDriver.Spec, cephv1.CephCOSIDriverSpec{}) && cephCOSIDriver.Spec.DeploymentStrategy != "" {
+		cosiDeploymentStrategy = cephCOSIDriver.Spec.DeploymentStrategy
+	}
+	if cosiDeploymentStrategy == "disabled" {
+		logger.Info("Ceph COSI Driver is disabled")
+		return reconcile.Result{}, *cephCOSIDriver, nil
+	}
+
+	r.recorder.Eventf(cephCOSIDriver, kapiv1.EventTypeNormal, "Reconciling", "Reconciling Ceph COSI Driver %q", cephCOSIDriver.Name)
+	// Set the default CephCOSIDriver name if not already set
+	if cephCOSIDriver.Name == "" {
+		cephCOSIDriver.Name = CephCOSIDriverName
+	}
+
+	// Set the default CephCOSIDriver namespace if not already set
+	if cephCOSIDriver.Namespace == "" {
+		cephCOSIDriver.Namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	}
+
+	// Check whether object store is running
+	list := &cephv1.CephObjectStoreList{}
+	err = r.client.List(r.opManagerContext, list)
+	if err != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to list object stores. %+v", err)
+	}
+
+	// TODO: Check for object store part of Multisite
+	if len(list.Items) == 0 && cosiDeploymentStrategy == "auto" {
+		logger.Info("no object stores found, skipping cosi driver config")
+		return waitForRequeueObjectStoreNotPresent, *cephCOSIDriver, nil
+	}
+
+	// TODO: create Ceph Object Store User name cosi-<objectstorename> with a secret containing its credentials for each object store, if not already created
+
+	// Start the Ceph COSI Driver
+	err = r.startCephCOSIDriver(cephCOSIDriver)
+	if err != nil {
+		return reconcile.Result{}, *cephCOSIDriver, fmt.Errorf("failed to start Ceph COSI Driver. %+v", err)
+	}
+
+	return reconcile.Result{}, *cephCOSIDriver, nil
+}
+
+// Start the Ceph COSI Driver
+func (r *ReconcileCephCOSIDriver) startCephCOSIDriver(cephCOSIDriver *cephv1.CephCOSIDriver) error {
+	// Create the Ceph COSI Driver Deployment
+	cephcosidriverDeployment, err := createCephCOSIDriverDeployment(cephCOSIDriver)
```

Comment:
Sorry I don't fully understand this part, Do I need to keep the creation of Deployment in the same function `startCephCOSIDriver` instead of helper `createCephCOSIDriverDeployment`

Reply:
Something must have been refactored already to fix this. I thought you were applying properties to the deployment in this method, but don't see it anymore, so you can ignore this.